AsyncDisplayKit/TextKit/ASTextKitFontSizeAdjuster.mm

```diff
@@ -149,53 +154,45 @@ - (CGFloat)scaleFactor
       }
     }
     
-    NSUInteger scaleIndex = 0;
-
-    // find the longest word and make sure it fits in the constrained width
-    if ([longestWordNeedingResize length] > 0) {
-      
-      NSRange longestWordRange = [str rangeOfString:longestWordNeedingResize];
-      NSMutableAttributedString *attrString = [textStorage attributedSubstringFromRange:longestWordRange].mutableCopy;
-      CGSize longestWordSize = [attrString boundingRectWithSize:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX) options:NSStringDrawingUsesLineFragmentOrigin context:nil].size;
-      
-      // check if the longest word is larger than our constrained width
-      if (longestWordSize.width > _constrainedSize.width) {
+    // check to see if we may need to shrink for any of these things
+    BOOL longestWordFits = [longestWordNeedingResize length] ? NO : YES;
+    BOOL maxLinesFits = _attributes.maximumNumberOfLines > 0 ? NO : YES;
+    BOOL heightFits = isinf(_constrainedSize.height) ? YES : NO;
+    
+    if (longestWordFits == NO || maxLinesFits == NO || heightFits == NO) {
+      // we may need to shrink for some reason, so let's iterate through our scale factors to see if we actually need to shrink
+      // Note: the first scale factor in the array is 1.0 so will make sure that things don't fit without shrinking
+      for (NSUInteger index = 0; index < scaleFactors.count && (longestWordFits == NO || maxLinesFits == NO || heightFits == NO); index++) {
         
-        // we have a word that is too long. Loop through our scale factors until we fit
-        for (NSNumber *scaleFactor in scaleFactors) {
-          // even if we still don't fit, save this scaleFactor so more of the word will fit
-          adjustedScale = [scaleFactor floatValue];
-          
-          // adjust here so we start at the proper place in our scale array if we have too many lines
-          scaleIndex++;
-          
-          if (std::ceil(longestWordSize.width * [scaleFactor floatValue])  <= _constrainedSize.width) {
-            // we fit! we are done
-            break;
-          }
+        adjustedScale = [scaleFactors[index] floatValue];
+        
+        if (longestWordFits == NO) {
+          // we need to check the longest word to make sure it fits
+          NSRange longestWordRange = [str rangeOfString:longestWordNeedingResize];
```

Comment:
Good call. I think we can move all of this except the multiplication out of the loop.